pilot/pkg/model/network_gw_names.go

```diff
@@ -0,0 +1,128 @@
+package model
+
+import (
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type nameCacheEntry struct {
+	value  []string
+	expiry time.Time
+}
+
+type networkGatewayNameCache struct {
+	sync.Mutex
+	client     *dnsClient
+	cache      map[string]nameCacheEntry
+	xdsUpdater XDSUpdater
+}
+
+func newNetworkGatewayNameCache(xdsUpdater XDSUpdater) (*networkGatewayNameCache, error) {
+	c, err := newClient()
+	if err != nil {
+		return nil, err
+	}
+	return &networkGatewayNameCache{client: c, xdsUpdater: xdsUpdater, cache: map[string]nameCacheEntry{}}, nil
+}
+
+func (n *networkGatewayNameCache) Resolve(name string) []string {
+	n.Lock()
+	defer n.Unlock()
+	if entry, ok := n.cache[name]; ok && entry.expiry.After(time.Now()) {
+		return entry.value
+	}
+	return n.resolveAndCache(name)
+}
+
+func (n *networkGatewayNameCache) resolveAndCache(name string) []string {
+	delete(n.cache, name)
+	names, ttl := n.resolve(name)
+	// don't cache for very small TTL or empty list of names
+	if len(names) == 0 || ttl < 10*time.Second {
+		return names
+	}
+	expiry := time.Now().Add(ttl)
+	time.AfterFunc(ttl, func() {
+		log.Infof("triggering update due to gateway hostname TTL expiry")
```

Comment:
should have more context on which one probably?

Can we only do it if the response is different as well?

Reply:
Also we should probably normalize the response order. Some DNS's will rotate them to get dumb clients to load balance; the order does not matter for envoy and will only cause update churn

Reply:
We sort the gateways later, no need to sort here.